backends/simplego/dotgeneral_fastpath.go

```diff
@@ -0,0 +1,283 @@
+// access pattern causes too many cache misses, and the normalized path (which
+// transposes RHS for sequential access) becomes faster despite the transpose overhead.
+//
+// This threshold was determined by benchmarking (BenchmarkDirectPathThreshold):
+//   - For [256, K] × [K, 256]: DirectPath wins at K≤128, NormalizedPath wins at K≥256
+//   - Crossover point is between K=128 and K=256
+//
+// Exception: For single-row operations (M=1), DirectPath is always faster because
+// the transpose overhead dominates when there's only one output row to compute.
+const DirectPathMaxContractingSize = 128
+
+// DirectPathMaxBatchSize is the maximum batch size for which the direct path is beneficial.
+// For larger batch sizes, the normalized path with batch parallelism is faster.
+// The direct path processes batches sequentially, while the normalized path can parallelize
+// across batches using multiple workers.
+const DirectPathMaxBatchSize = 64
+
+// canUseDirectPath determines if we can use the direct (no-transpose) execution path.
+//
+// The direct path skips normalization/transpose but has strided RHS access.
+// It's beneficial when:
+//  1. The dtype is float32 (currently the only optimized implementation)
+//  2. The axes are already in contract-last order for LHS
+//  3. The batch size is small (direct path doesn't parallelize across batches)
+//  4. Either:
+//     a. Single-row operation (lhsCrossSize=1) where transpose overhead dominates, OR
+//     b. Small contracting dimension where strided access doesn't cause excessive cache misses
+//
+// For larger matrices or batch sizes, use execDotGeneralSmallNormalized or execDotGeneralBlocked instead.
+func canUseDirectPath(lhs, rhs *Buffer, params *dotGeneralNodeData) bool {
+	// Only support float32 direct path for now (most common)
+	if lhs.shape.DType != dtypes.Float32 {
+		return false
+	}
+
+	// Check if axes are in contract-last order (standard matmul layout)
+	if !isContractLastOrder(lhs.shape, rhs.shape,
+		params.lhsContractingAxes, params.rhsContractingAxes,
+		params.lhsBatchAxes, params.rhsBatchAxes) {
+		return false
+	}
+
+	// For large batch sizes, the normalized path with batch parallelism is faster.
+	// The direct path processes batches sequentially without parallelization.
+	if params.batchSize > DirectPathMaxBatchSize {
+		return false
+	}
+
+	// For single-row operations (M=1) with small batch sizes, direct path is faster
+	// because transpose overhead dominates when computing just one output row per batch.
+	// Benchmarks show DirectPath is 10-15x faster for M=1, batchSize=1 cases.
+	// Note: Large batch sizes are handled above (parallelization wins).
+	if params.lhsCrossSize == 1 {
+		return true
+	}
+
+	// For multi-row operations, only use direct path when contracting dimension
+	// is small enough that strided RHS access doesn't cause excessive cache misses.
+	if params.contractingSize > DirectPathMaxContractingSize {
```

Comment:
See my comments above: my intuition is that RHS.cross sizes matter more than the contracting size. Maybe you know better ?